### PR TITLE
types,executor: treat decimal truncate as warning in update

### PR DESCRIPTION
**Pull request by tiancaiamao**
Fix issue https://github.com/pingcap/tidb/issues/5532

@coocood  @zimulala  @XuHuaiyu

### CODE REVIEW

File: executor/write_test.go [hunk 1]
```diff
@@ -648,6 +648,14 @@ func (s *testSuite) TestUpdate(c *C) {
...
 	r2 := tk.MustQuery("select ts from tsup;")
 	r1.Check(r2.Rows())
+
+	// issue 5532
+	tk.MustExec("create table decimals (a decimal(20, 0) not null)")
+	tk.MustExec("insert into decimals values (201)")
+	// A warning rather than data truncated error.
+	tk.MustExec("update decimals set a = a + 1.23;")
+	r = tk.MustQuery("select * from decimals")
+	r.Check(testkit.Rows("202"))
```
Comment (XuHuaiyu) [inline executor/write_test.go]: add show warnings here

File: executor/statement_context_test.go [hunk 2]
```diff
@@ -37,9 +36,7 @@ func (s *testSuite) TestStatementContext(c *C) {
 
 	tk.MustExec(strictModeSQL)
 	tk.MustQuery("select * from sc where a > cast(1.1 as decimal)").Check(testkit.Rows("2"))
-	_, err := tk.Exec(`select * from sc where a > cast(1.1 as decimal);
-		        update sc set a = 4 where a > cast(1.1 as decimal)`)
-	c.Check(terror.ErrorEqual(err, types.ErrTruncated), IsTrue)
+	tk.MustExec("update sc set a = 4 where a > cast(1.1 as decimal)")
```
Comment (tiancaiamao) [inline executor/statement_context_test.go]: I've tested that this behavior confirms to MySQL 5.6, 5.7...
The previous code is wrong.

**Comment by coocood** [comment]
LGTM

File: types/datum.go [hunk 3]
```diff
@@ -1087,8 +1087,9 @@ func ProduceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *stmtctx.Statem
 				return nil, errors.Trace(err)
 			}
 			if !dec.IsZero() && frac > decimal && dec.Compare(&old) != 0 {
-				if sc.InInsertStmt {
+				if sc.InInsertStmt || sc.InUpdateOrDeleteStmt {
```
Comment (zimulala) [inline types/datum.go]: Do we need to add a test for `delete` SQL?

Comment (tiancaiamao) [reply types/datum.go]: Delete behavior is different from MySQL:

```
mysql> select * from decimals;
+-----+
| a   |
+-----+
| 201 |
+-----+
1 row in set (0.00 sec)
```

MySQL:

```
mysql> delete from decimals where a =  cast(201.2 as decimal(20,0));
Query OK, 1 row affected (0.00 sec)
```

TiDB:

```
mysql> delete from decimals where a =  cast(201.2 as decimal(20,0));
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------+
| Level   | Code | Message        |
+---------+------+----------------+
| Warning | 1265 | Data Truncated |
+---------+------+----------------+
1 row in set (0.00 sec)
```

Although it's not the same behavior as MySQL, the result in this PR is more acceptable for delete statement. 
@zimulala

**Pull request merged**